### DISCUSSION TITLE
Trouble starting with including vips in C++ code

### DISCUSSION THREAD

**Discussion by patkalie** [libvips/libvips#2105] [Help wanted]
Hi,

I am make a first try to use vips from C++ code, and I failed at an very early stage. 

I tried to strat out by using the sample code "invert.cpp" found here.
My compiler fails at "#include <vips/vips8>" when I try to build, and it says: "fatal error: vips/vips8: No such file or directory".

The vips-folder containing the vips8-file is present in the same folder as my cpp-file with the code copied from the "invert.cpp".

I guess I am doing some very obvious mistakes here. What will I have to to besides including the vips-folder in my my c++ project root folder. Can I include all the Libvips cpp-files and h-files in my project? Is it necessary to install Libvips or other tools on beforehand? 

I would very much appreciate it if somebody could point me to a place where I can read how to include the necessary files to make the "invert.cpp" example code compile.

Kind regards,
Kai-Inge

**Comment by jcupitt** [comment]
Hi @patkalie, it depends on your platform. For linux, just install libvips with your package manager, then enter something like:

```
g++ -g -Wall my-program.cpp $(pkg-config vips-cpp --cflags --libs)
```

**Reply by patkalie** [reply]
Hi @jcupitt, thanks for your quick answer. "Unfortunately" I am using Windows 10. I have tried using the above command in my computer but the "pkg-config" is not recognized. I guess I need use something like "-I C:\mylibs\vips-dev-8.10" as part of the g++-command?

**Reply by jcupitt** [reply]
It's more complex. You'll need to build the C++ layer of libvips yourself.

Could you use WSL2? That would let you use the linux build environment on your Windows PC.

**Reply by patkalie** [reply]
Thank you! I will have to read up on the WSL2 option. If I choose to develop in C on my Win 10 computer, will it be just as complex as using C++?

**Reply by jcupitt** [reply]
C is much simpler.

You can't mix compilers in C++ projects, so you need to rebuild the C++ parts of the libvips binary before you can use it with MSVC. But all compilers support the same C linkage, so you can call into libvips.dll from any compiler.

Just work though the dialog boxes putting the correct -I and -L paths in where necessary, like any external library.

**Reply by patkalie** [reply]
Again, thanks a lot  @jcupitt for your time. I have just started learning  C and C++ and I wanted to try to include vips in my "learning-by-coding small useful stuff". I have not yet advanced to the topic of accessing dlls from my C-code but I look forward to it. Have a nice day! Kai
